lib/mixlib/shellout.rb

```diff
@@ -199,9 +202,48 @@ def uid
 
     # The gid that the subprocess will switch to. If the group attribute is
     # given as a group name, it is converted to a gid by Etc.getgrnam
+    # If a group is not provided but with_logon is true, the user's primary
+    # group will be used.
     def gid
-      return nil unless group
-      group.kind_of?(Integer) ? group : Etc.getgrnam(group.to_s).gid
+      return group.kind_of?(Integer) ? group : Etc.getgrnam(group.to_s).gid if group 
+      return Etc.getpwuid(uid).gid if (with_logon && user)
+      return nil
+    end
+
+    # Helper method for sgids
+    def all_seconderies
+      ret = []
+      Etc.endgrent
+      while ( g = Etc.getgrent ) do
+        ret << g
+      end
+      Etc.endgrent
+      return ret
+    end
+
+    # The secondary groups that the subprocess will switch to.
+    # Currently valid only if with_logon is used, and is set
+    # to the user's secondary groups
+    def sgids
+      return nil unless with_logon && user
+      user_name = Etc.getpwuid(uid).name
+      all_seconderies.select{|g| g.mem.include?(user_name)}.map{|g|g.gid}
+    end
+
+    # The environment variables that are deduced from simulating logon
+    # Only valid if with_logon is used
+    def logon_environment
+      return {} unless (with_logon && user)
+      entry = Etc.getpwuid(uid)
+      # According to `man su`, the set fields are:
+      #  $HOME, $SHELL, $USER, $LOGNAME, $PATH, and $IFS 
+      # Values are copied from "shadow" package in Ubuntu 14.10
+      {'HOME'=>entry.dir, 'SHELL'=>entry.shell, 'USER'=>entry.name, 'LOGNAME'=>entry.name, 'PATH'=>'/sbin:/bin:/usr/sbin:/usr/bin', 'IFS'=>"\t\n"}
+    end
+
+    # Merges the two environments for the process
+    def process_environment
+      logon_environment.merge(self.environment)
     end
```

Comment:
All the stuff you added I'm pretty certain should go into unix.rb since its all unix-specific.

Grepping through the sources I only see gid used on the unix side, so I think its safe to move that into unix.rb as well.